`WCOrderStatusModel` to `Room`

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

Closes: AINFRA-557

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCOrderStatusModel` from `WellSql` to `Room` database.

### Testing information
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

#### Standard statuses
1. Open the app
2. Create an order
3. Edit the order: see you can see all default statuses: `Cancelled`, `Draft`, `Completed`, `Failed`, `On hold`, `Pending payment`, `Processing`.

#### Custom status
1. On your WooCommerce web instance, go to `Tools` > `Theme file editor` > `Theme functions`
2. Add a snippet that adds a custom status (available below)
3. Open the mobile app
4. See that the custom status is available when you edit an order
5. Set the custom status 
6. See that the custom status was set on web as well
7. On the app, change the status back to a standard status (e.g. `processing`)
8. Remove the snippet added in point 2
9. Restart the mobile app
10. See that the custom status is no longer available in the mobile app.

<details><summary>Snippet to add a custom order status</summary>

```php
add_action('init', function () {
    register_post_status('wc-custom', [
        'label'                     => 'Custom status',
        'public'                    => true,
        'exclude_from_search'       => false,
        'show_in_admin_all_list'    => true,
        'show_in_admin_status_list' => true,
        'label_count'               => _n_noop('Custom status <span class="count">(%s)</span>', 'Custom status <span class="count">(%s)</span>')
    ]);
});

add_filter('wc_order_statuses', function ($order_statuses) {
    return array_merge(
        array_slice($order_statuses, 0, array_search('wc-processing', array_keys($order_statuses)) + 1),
        ['wc-custom' => 'Custom status'],
        array_slice($order_statuses, array_search('wc-processing', array_keys($order_statuses)) + 1)
    );
});

```

</details> 

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
As above

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="927" height="432" alt="Screenshot 2025-07-30 at 12 26 31" src="https://github.com/user-attachments/assets/171544aa-04c9-4d63-a864-77fa824acc02" />


- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Migrate WCOrderStatusModel to Room

Introduce Dao, provide migrations, adjust the code to changes signatures

wzieba committed:
Remove unused methods from OrderSqlUtils

wzieba committed:
Remove `rowsAffected` field from `OnOrderStatusOptionsChanged`

It was not used anywhere

wzieba committed:
Remove `runBlocking` from `handleFetchOrderStatusOptionsCompleted`

wzieba committed:
Merge branch 'migrate_order_summary' into migrate_status_model

wzieba committed:
Fix `WooCommerce` test execution

But with 2 `OrderCreateEditRepositoryTest` tests failing

wzieba committed:
Merge remote-tracking branch 'origin/trunk' into migrate_status_model

wzieba committed:
Fix `OrderCreateEditRepositoryTest` tests

They worked before as `WCOrderStatusModel` didn't have a correct implementation of `equals` method. So in the end, we were comparing `WCOrderStatusModel` excluding `statusKey` or `label`.

The behavior was not changed by the migration, the test only got more strict.

wzieba committed:
Apply formatting changes

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Remove outdated unit tests

wzieba committed:
Add tests for `OrderStatusDao`

wzieba committed:
Fix formatting

wzieba committed:
Rename `localSiteId` to `siteId`

wzieba committed:
Rename `localSiteId` to `siteId`

wzieba committed:
Introduce `OrderStatusDao#replaceAll` method

I couldn't identify a reason for keeping overcomplex logic of keeping statuses up to date by adding/deleting single statuses.

wzieba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCOrderStatusModel.kt:
```diff
@@ -1,25 +1,15 @@
+    val siteId: LocalId = LocalId(0),
+    val statusKey: String = "",
+    val label: String = "",
+    val statusCount: Int = 0
```
I leave the default arguments as in some parts of the code, `WCOrderStatusModel` is treated as a kind of wrapper over the string (e.g. `WCOrderStatusModel(statusKey = AUTO_DRAFT)` ) and I didn't want to modify this flow (at least now).

wzieba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt:
```diff
@@ -1103,69 +1103,24 @@ class WCOrderStore @Inject internal constructor(
-    }
-
-    @Suppress("NestedBlockDepth")
-    private fun addOrUpdateOptions(
-        payload: FetchOrderStatusOptionsResponsePayload,
-        existingOptions: List<WCOrderStatusModel>
-    ): List<WCOrderStatusModel> {
-        val addOrUpdateOptions = mutableListOf<WCOrderStatusModel>()
-        payload.labels.iterator().forEach { newOption ->
-            var exists = false
-            existingOptions.iterator().forEach eoi@{ existingOption ->
-                if (newOption.statusKey == existingOption.statusKey) {
-                    exists = true
-                    if (newOption.label != existingOption.label ||
-                        newOption.statusCount != existingOption.statusCount) {
-                        addOrUpdateOptions.add(newOption)
-                    }
-                    return@eoi
-                }
-            }
-            if (!exists) addOrUpdateOptions.add(newOption)
-        }
-        return addOrUpdateOptions
-    }
-
-    private fun deleteOptions(
-        payload: FetchOrderStatusOptionsResponsePayload,
-        existingOptions: List<WCOrderStatusModel>
-    ): List<WCOrderStatusModel> {
-        val deleteOptions = mutableListOf<WCOrderStatusModel>()
-        existingOptions.iterator().forEach { existingOption ->
-            var exists = false
-            payload.labels.iterator().forEach noi@{ newOption ->
-                if (newOption.statusKey == existingOption.statusKey) {
-                    exists = true
-                    return@noi
-                }
-            }
-            if (!exists) deleteOptions.add(existingOption)
-        }
```
I couldn't find a justification for this complex logic, so instead I propose a simply `replaceAll` method. Maybe it was needed for getting `rowsAffected` value, but this property is not used anymore.

ParaskP7 replied:
Yea @wzieba , I searched/thought about it too, and all I got is this ebe29139d4b5047045098fb2372318eedbc7acfb commit, which didn't help me much, as I couldn't (fully) understand the reasoning from the commit's description... 🤷

wzieba committed:
Remove unused `WellSql` configuration from `WCOrderStoreTest`

wzieba committed:
Remove unused test resources

wzieba committed:
Remove unused imports

wzieba commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditRepositoryTest.kt:
```diff
@@ -200,7 +200,7 @@ class OrderCreateEditRepositoryTest : BaseUnitTest() {
         // THEN the order status is not changed
         val request = UpdateOrderRequest(
-            status = WCOrderStatusModel(Order.Status.AUTO_DRAFT),
+            status = WCOrderStatusModel(statusKey = Order.Status.Pending.value, label = Order.Status.Pending.value),
```
Generally a good comment, but this is explained in commit message: 3ad15d7358f65bc55088b78c74b09531a3d061ae

wzieba commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditRepositoryTest.kt:
```diff
@@ -228,7 +228,7 @@ class OrderCreateEditRepositoryTest : BaseUnitTest() {
         // Then the order status is not changed
         val request = UpdateOrderRequest(
-            status = WCOrderStatusModel(Order.Status.AUTO_DRAFT),
+            status = WCOrderStatusModel(statusKey = Order.Status.Pending.value, label = Order.Status.Pending.value),
```
https://github.com/woocommerce/woocommerce-android/pull/14373#discussion_r2242383322

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/data/ObserveSiteOrdersState.kt:
```diff
@@ -45,7 +45,7 @@ class ObserveSiteOrdersState @Inject constructor(
-            .onStart { emit(OnOrderStatusOptionsChanged(0)) }
+            .onStart { emit(OnOrderStatusOptionsChanged()) }
             .map {
                 orderStore.getOrderStatusOptionsForSite(site)
```
Not really, the context is suspend. Otherwise we would have a compilation error

<img width="667" height="259" alt="image" src="https://github.com/user-attachments/assets/a16e96d0-263b-4564-a160-02d41d4b5443" />

wzieba commented on libs/fluxc-plugin/src/test/java/org/wordpress/android/fluxc/store/OrderUpdateStoreTest.kt:
```diff
@@ -543,7 +543,7 @@ class OrderUpdateStoreTest {
         const val SIMPLE_PAYMENT_BILLING_EMAIL = "example@example.com"
         const val SIMPLE_PAYMENT_IS_TAXABLE = true
-        val SIMPLE_PAYMENT_ORDER_STATUS = WCOrderStatusModel(CoreOrderStatus.PROCESSING.value)
+        val SIMPLE_PAYMENT_ORDER_STATUS = WCOrderStatusModel(label = CoreOrderStatus.PROCESSING.value)
```
I've seen this, but intentionally I didn't change it:
- all tests still pass, so it seems `label` is not used at all
- `statusKey` is more data-driven property (`label` is UI) so I think setting this in context of `OrderUpdateStoreTest` is more appropriate

wzieba committed:
Drop `WCOrderStatusModel` from WellSql

ParaskP7 reviewed (approved):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything works as expected (ish), one more migration done! 🌟 

---

I have zero comment for you this time, but I noticed something while testing, might be important (or not so much). Maybe something to ask of the product teams before merging: 🤷

Before the change, when clicking on edit to change the order status, the order of statuses in that `Change order status` was:

<img width="270" height="600" alt="before" src="https://github.com/user-attachments/assets/28604108-af3a-4f32-895d-fa66584404bd" />

After the change, it now looks like:

<img width="270" height="600" alt="after" src="https://github.com/user-attachments/assets/61bfd41e-939b-4306-94fa-907f5baaad38" />

My question is whether this is acceptable, or else whether it will create some confusion to our users, which by now are used to the "before" order of statuses.

Wdyt? 🤔

wzieba commented:
Nice find, thanks @ParaskP7 ! I guess we want to keep the previous behavior which is in sync with web version

<img width="315" height="400" alt="Screenshot 2025-07-31 at 13 47 16" src="https://github.com/user-attachments/assets/a1132c23-681b-4341-8012-ffb1f308cf34" />

I'll think how to address it

ParaskP7 commented:
> I guess we want to keep the previous behavior which is in sync with web version

I agree @wzieba ! 💯 

> I'll think how to address it

Thanks! 👍

wzieba committed:
Add `WCOrderStatusModel#position`

This will assert the same order as when fetched from the API (which is expected, and in sync with wp-admin UI)

wzieba commented:
I've addressed this by adding a new field: `position` 8a21ec2. There's no anything we could use from API directly ([link](https://woocommerce.github.io/woocommerce-rest-api-docs/#retrieve-orders-totals)) so relying on order of statuses when fetching sounds like a good solution to me.

<img width="250" height="1123" alt="Screenshot 2025-07-31 at 14 36 31" src="https://github.com/user-attachments/assets/238c3dbf-6184-4812-b647-76c3cf416681" />

ParaskP7 commented:
> I've addressed this by adding a new field: position https://github.com/woocommerce/woocommerce-android/commit/8a21ec20cbe4d3435ee9df218b9893eae431216a. There's no anything we could use from API directly ([link](https://woocommerce.github.io/woocommerce-rest-api-docs/#retrieve-orders-totals)) so relying on order of statuses when fetching sounds like a good solution to me.

This is great @wzieba , thank you for the quick fix and that nifty trick you did with `mapIndexed` to get the `position` for us (for free), looks great now, nice! 🚀
